int.ts

```diff
@@ -714,59 +703,66 @@ module J2ME {
+            arrayAddr = ref[--sp];
+            if ((index >>> 0) >= (i32[arrayAddr >> 2] >>> 0)) {
               thread.throwException(fp, sp, opPC, ExceptionType.ArrayIndexOutOfBoundsException, index);
             }
-            i32[sp++] = array[index];
+            i32[sp++] = i8[arrayAddr + 4 + index];
             continue;
           case Bytecodes.CALOAD:
             index = i32[--sp];
-            array = ref[--sp];
-            if ((index >>> 0) >= (array.length >>> 0)) {
+            arrayAddr = ref[--sp];
+            if ((index >>> 0) >= (i32[arrayAddr >> 2] >>> 0)) {
               thread.throwException(fp, sp, opPC, ExceptionType.ArrayIndexOutOfBoundsException, index);
             }
-            i32[sp++] = array[index];
+            i32[sp++] = u16[arrayAddr/2 + 2 + index];
             continue;
           case Bytecodes.SALOAD:
             index = i32[--sp];
-            array = ref[--sp];
-            if ((index >>> 0) >= (array.length >>> 0)) {
+            arrayAddr = ref[--sp];
+            if ((index >>> 0) >= (i32[arrayAddr >> 2] >>> 0)) {
               thread.throwException(fp, sp, opPC, ExceptionType.ArrayIndexOutOfBoundsException, index);
             }
-            i32[sp++] = array[index];
+            i32[sp++] = i16[(arrayAddr >> 1) + 2 + index];
             continue;
           case Bytecodes.FALOAD:
             index = i32[--sp];
-            array = ref[--sp];
-            if ((index >>> 0) >= (array.length >>> 0)) {
+            arrayAddr = ref[--sp];
+            if ((index >>> 0) >= (i32[arrayAddr >> 2] >>> 0)) {
               thread.throwException(fp, sp, opPC, ExceptionType.ArrayIndexOutOfBoundsException, index);
             }
-            f32[sp++] = array[index];
+            f32[sp++] = f32[(arrayAddr >> 2) + 1 + index];
             continue;
           case Bytecodes.AALOAD:
             index = i32[--sp];
-            array = ref[--sp];
-            if ((index >>> 0) >= (array.length >>> 0)) {
+            arrayAddr = ref[--sp];
+
+            if (arrayAddr === 0) {
+              thread.throwException(fp, sp, opPC, ExceptionType.NullPointerException);
+              continue;
+            }
+
+            if ((index >>> 0) >= (i32[arrayAddr >> 2] >>> 0)) {
               thread.throwException(fp, sp, opPC, ExceptionType.ArrayIndexOutOfBoundsException, index);
             }
-            ref[sp++] = array[index];
+
+            ref[sp++] = i32[(arrayAddr >> 2) + 1 + index];
             continue;
           case Bytecodes.DALOAD:
             index = i32[--sp];
-            array = ref[--sp];
-            if ((index >>> 0) >= (array.length >>> 0)) {
+            arrayAddr = ref[--sp];
+            if ((index >>> 0) >= (i32[arrayAddr >> 2] >>> 0)) {
               thread.throwException(fp, sp, opPC, ExceptionType.ArrayIndexOutOfBoundsException, index);
             }
-            aliasedF64[0] = array[index];
+            aliasedF64[0] = getArrayFromAddr(arrayAddr)[index];
```

Comment:
Can we get the value directly from memory instead of indirecting through *getArrayFromAddr*? In this case we know (I think) that *arrayAddr* is an address.